"SQLException: Not using JDBC" when trying to obtain lock for entity created in same transaction

When trying to obtain a pessimistic lock for an entity that was created within the same transaction that is used to obtain the lock, Hibernate Reactive fails with the following exception:

```
2022-04-07 11:39:09,398 ERROR [org.jbo.res.rea.com.cor.AbstractResteasyReactiveContext] (vert.x-eventloop-thread-2) Request failed: org.hibernate.dialect.lock.PessimisticEntityLockException: could not obtain pessimistic lock
	at org.hibernate.dialect.lock.PessimisticWriteSelectLockingStrategy.lock(PessimisticWriteSelectLockingStrategy.java:102)
	at org.hibernate.persister.entity.AbstractEntityPersister.lock(AbstractEntityPersister.java:1996)
	at org.hibernate.event.internal.AbstractLockUpgradeEventListener.upgradeLock(AbstractLockUpgradeEventListener.java:82)
	at org.hibernate.loader.entity.CacheEntityLoaderHelper.loadFromSessionCache(CacheEntityLoaderHelper.java:111)
	at org.hibernate.reactive.event.impl.DefaultReactiveLoadEventListener.doLoad(DefaultReactiveLoadEventListener.java:597)
	at org.hibernate.reactive.event.impl.DefaultReactiveLoadEventListener.load(DefaultReactiveLoadEventListener.java:289)
	at org.hibernate.reactive.event.impl.DefaultReactiveLoadEventListener.lockAndLoad(DefaultReactiveLoadEventListener.java:550)
	at org.hibernate.reactive.event.impl.DefaultReactiveLoadEventListener.doOnLoad(DefaultReactiveLoadEventListener.java:213)
	at org.hibernate.reactive.event.impl.DefaultReactiveLoadEventListener.lambda$reactiveOnLoad$1(DefaultReactiveLoadEventListener.java:140)
	at java.base/java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:1106)
	at java.base/java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2235)
	at java.base/java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:143)
	at org.hibernate.reactive.event.impl.DefaultReactiveLoadEventListener.reactiveOnLoad(DefaultReactiveLoadEventListener.java:139)
	at org.hibernate.event.service.internal.EventListenerGroupImpl.lambda$fireEventOnEachListener$1(EventListenerGroupImpl.java:148)
	at java.base/java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:1106)
	at java.base/java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2235)
	at java.base/java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:143)
	at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:148)
	at org.hibernate.reactive.session.impl.ReactiveSessionImpl.fireLoadNoChecks(ReactiveSessionImpl.java:1084)
	at org.hibernate.reactive.session.impl.ReactiveSessionImpl.fireLoad(ReactiveSessionImpl.java:1077)
	at org.hibernate.reactive.session.impl.ReactiveSessionImpl$ReactiveIdentifierLoadAccessImpl.doLoad(ReactiveSessionImpl.java:1227)
	at org.hibernate.reactive.session.impl.ReactiveSessionImpl$ReactiveIdentifierLoadAccessImpl.lambda$load$5(ReactiveSessionImpl.java:1213)
	at org.hibernate.reactive.session.impl.ReactiveSessionImpl$ReactiveIdentifierLoadAccessImpl.perform(ReactiveSessionImpl.java:1180)
	at org.hibernate.reactive.session.impl.ReactiveSessionImpl$ReactiveIdentifierLoadAccessImpl.load(ReactiveSessionImpl.java:1213)
	at org.hibernate.reactive.session.impl.ReactiveSessionImpl.reactiveFind(ReactiveSessionImpl.java:1020)
	at org.hibernate.reactive.mutiny.impl.MutinySessionImpl.lambda$find$6(MutinySessionImpl.java:124)
	at io.smallrye.context.impl.wrappers.SlowContextualSupplier.get(SlowContextualSupplier.java:21)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateFromCompletionStage.subscribe(UniCreateFromCompletionStage.java:24)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.operators.uni.UniRunSubscribeOn.lambda$subscribe$0(UniRunSubscribeOn.java:27)
	at org.hibernate.reactive.context.impl.VertxContext.execute(VertxContext.java:78)
	at io.smallrye.mutiny.operators.uni.UniRunSubscribeOn.subscribe(UniRunSubscribeOn.java:25)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransform.subscribe(UniOnItemTransform.java:22)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransformToUni$UniOnItemTransformToUniProcessor.performInnerSubscription(UniOnItemTransformToUni.java:81)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransformToUni$UniOnItemTransformToUniProcessor.onItem(UniOnItemTransformToUni.java:57)
	at io.smallrye.mutiny.operators.uni.UniOperatorProcessor.onItem(UniOperatorProcessor.java:46)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateFromCompletionStage$CompletionStageUniSubscription.forwardResult(UniCreateFromCompletionStage.java:63)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenCompleteStage(CompletableFuture.java:883)
	at java.base/java.util.concurrent.CompletableFuture.whenComplete(CompletableFuture.java:2251)
	at java.base/java.util.concurrent.CompletableFuture.whenComplete(CompletableFuture.java:143)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateFromCompletionStage$CompletionStageUniSubscription.forward(UniCreateFromCompletionStage.java:51)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateFromCompletionStage.subscribe(UniCreateFromCompletionStage.java:35)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.operators.uni.UniRunSubscribeOn.lambda$subscribe$0(UniRunSubscribeOn.java:27)
	at org.hibernate.reactive.context.impl.VertxContext.execute(VertxContext.java:78)
	at io.smallrye.mutiny.operators.uni.UniRunSubscribeOn.subscribe(UniRunSubscribeOn.java:25)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransformToUni.subscribe(UniOnItemTransformToUni.java:25)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransformToUni$UniOnItemTransformToUniProcessor.performInnerSubscription(UniOnItemTransformToUni.java:81)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransformToUni$UniOnItemTransformToUniProcessor.onItem(UniOnItemTransformToUni.java:57)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateFromCompletionStage$CompletionStageUniSubscription.forwardResult(UniCreateFromCompletionStage.java:63)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2073)
	at io.vertx.core.Future.lambda$toCompletionStage$2(Future.java:360)
	at io.vertx.core.impl.future.FutureImpl$3.onSuccess(FutureImpl.java:141)
	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:60)
	at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:211)
	at io.vertx.core.impl.future.PromiseImpl.tryComplete(PromiseImpl.java:23)
	at io.vertx.core.impl.future.PromiseImpl.onSuccess(PromiseImpl.java:49)
	at io.vertx.core.impl.future.PromiseImpl.handle(PromiseImpl.java:41)
	at io.vertx.sqlclient.impl.TransactionImpl.lambda$wrap$0(TransactionImpl.java:72)
	at io.vertx.core.impl.future.FutureImpl$3.onSuccess(FutureImpl.java:141)
	at io.vertx.core.impl.future.FutureBase.lambda$emitSuccess$0(FutureBase.java:54)
	at io.vertx.core.impl.EventLoopContext.execute(EventLoopContext.java:81)
	at io.vertx.core.impl.DuplicatedContext.execute(DuplicatedContext.java:173)
	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:51)
	at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:211)
	at io.vertx.core.impl.future.PromiseImpl.tryComplete(PromiseImpl.java:23)
	at io.vertx.core.impl.future.PromiseImpl.onSuccess(PromiseImpl.java:49)
	at io.vertx.core.impl.future.PromiseImpl.handle(PromiseImpl.java:41)
	at io.vertx.core.impl.future.PromiseImpl.handle(PromiseImpl.java:23)
	at io.vertx.pgclient.impl.PgSocketConnection.lambda$doSchedule$2(PgSocketConnection.java:160)
	at io.vertx.sqlclient.impl.command.CommandResponse.fire(CommandResponse.java:46)
	at io.vertx.sqlclient.impl.SocketConnectionBase.handleMessage(SocketConnectionBase.java:287)
	at io.vertx.pgclient.impl.PgSocketConnection.handleMessage(PgSocketConnection.java:96)
	at io.vertx.sqlclient.impl.SocketConnectionBase.lambda$init$0(SocketConnectionBase.java:99)
	at io.vertx.core.impl.EventLoopContext.emit(EventLoopContext.java:50)
	at io.vertx.core.impl.ContextImpl.emit(ContextImpl.java:274)
	at io.vertx.core.impl.EventLoopContext.emit(EventLoopContext.java:22)
	at io.vertx.core.net.impl.NetSocketImpl.handleMessage(NetSocketImpl.java:394)
	at io.vertx.core.net.impl.ConnectionBase.read(ConnectionBase.java:156)
	at io.vertx.core.net.impl.VertxHandler.channelRead(VertxHandler.java:153)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436)
	at io.vertx.pgclient.impl.codec.PgEncoder.lambda$write$0(PgEncoder.java:87)
	at io.vertx.pgclient.impl.codec.PgCommandCodec.handleReadyForQuery(PgCommandCodec.java:139)
	at io.vertx.pgclient.impl.codec.PgDecoder.decodeReadyForQuery(PgDecoder.java:237)
	at io.vertx.pgclient.impl.codec.PgDecoder.channelRead(PgDecoder.java:96)
	at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:251)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:722)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:658)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:584)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:496)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: org.hibernate.exception.GenericJDBCException: Unable to acquire JDBC Connection
	at org.hibernate.exception.internal.StandardSQLExceptionConverter.convert(StandardSQLExceptionConverter.java:42)
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:113)
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:99)
	at org.hibernate.resource.jdbc.internal.LogicalConnectionManagedImpl.acquireConnectionIfNeeded(LogicalConnectionManagedImpl.java:111)
	at org.hibernate.resource.jdbc.internal.LogicalConnectionManagedImpl.getPhysicalConnection(LogicalConnectionManagedImpl.java:138)
	at org.hibernate.engine.jdbc.internal.StatementPreparerImpl.connection(StatementPreparerImpl.java:50)
	at org.hibernate.engine.jdbc.internal.StatementPreparerImpl$1.doPrepare(StatementPreparerImpl.java:90)
	at org.hibernate.engine.jdbc.internal.StatementPreparerImpl$StatementPreparationTemplate.prepareStatement(StatementPreparerImpl.java:176)
	at org.hibernate.engine.jdbc.internal.StatementPreparerImpl.prepareStatement(StatementPreparerImpl.java:75)
	at org.hibernate.dialect.lock.PessimisticWriteSelectLockingStrategy.lock(PessimisticWriteSelectLockingStrategy.java:62)
	... 112 more
Caused by: java.sql.SQLException: Not using JDBC
	at org.hibernate.reactive.provider.service.NoJdbcConnectionProvider.getConnection(NoJdbcConnectionProvider.java:25)
	at org.hibernate.internal.NonContextualJdbcConnectionAccess.obtainConnection(NonContextualJdbcConnectionAccess.java:38)
	at org.hibernate.resource.jdbc.internal.LogicalConnectionManagedImpl.acquireConnectionIfNeeded(LogicalConnectionManagedImpl.java:108)
	... 118 more
```

Of course, locking an entity while it only exists in the local transaction isn't really necessary. But it would still be nice if Hibernate Reactive supported this use case. For example, we sometimes want to update an entity and don't want to distinguish whether the entity was just created earlier in the transaction or whether it already existed. So in the current situation we would have to distinguish which use case is executed and either access the entity with lock or without lock. If not for this bug, we could just reuse the update code for multiple use cases and have the exact same behavior.

Reproducer: https://github.com/markusdlugi/hibernate-locking

Steps to reproduce:

1. Start the attached MyEntityResourceTest.
2. Observe that creating an entity and then updating it with lock leads to the aforementioned exception, while updating it without lock works.

Tested with Quarkus 2.7.5.Final and Hibernate Reactive 1.1.3.Final.

--- a/hibernate-reactive-core/src/main/java/org/hibernate/reactive/event/impl/DefaultReactiveLoadEventListener.java
+++ b/hibernate-reactive-core/src/main/java/org/hibernate/reactive/event/impl/DefaultReactiveLoadEventListener.java
@@ -32,6 +32,7 @@ import org.hibernate.persister.entity.EntityPersister;
 import org.hibernate.proxy.HibernateProxy;
 import org.hibernate.proxy.LazyInitializer;
 import org.hibernate.reactive.event.ReactiveLoadEventListener;
+import org.hibernate.reactive.loader.entity.ReactiveCacheEntityLoaderHelper;
 import org.hibernate.reactive.logging.impl.Log;
 import org.hibernate.reactive.logging.impl.LoggerFactory;
 import org.hibernate.reactive.persister.entity.impl.ReactiveEntityPersister;
@@ -594,7 +595,7 @@ public class DefaultReactiveLoadEventListener implements LoadEventListener, Reac
 		}
 
 		CacheEntityLoaderHelper.PersistenceContextEntry persistenceContextEntry =
-				CacheEntityLoaderHelper.INSTANCE.loadFromSessionCache(
+				ReactiveCacheEntityLoaderHelper.loadFromSessionCache(
 						event,
 						keyToLoad,
 						options

--- a/hibernate-reactive-core/src/main/java/org/hibernate/reactive/loader/entity/ReactiveCacheEntityLoaderHelper.java
+++ b/hibernate-reactive-core/src/main/java/org/hibernate/reactive/loader/entity/ReactiveCacheEntityLoaderHelper.java
@@ -1,0 +1,143 @@
+/* Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * Copyright: Red Hat Inc. and Hibernate Authors
+ */
+package org.hibernate.reactive.loader.entity;
+
+import org.hibernate.HibernateException;
+import org.hibernate.LockMode;
+import org.hibernate.LockOptions;
+import org.hibernate.ObjectDeletedException;
+import org.hibernate.cache.spi.access.EntityDataAccess;
+import org.hibernate.cache.spi.access.SoftLock;
+import org.hibernate.engine.spi.EntityEntry;
+import org.hibernate.engine.spi.EntityKey;
+import org.hibernate.engine.spi.SessionImplementor;
+import org.hibernate.engine.spi.Status;
+import org.hibernate.event.spi.EventSource;
+import org.hibernate.event.spi.LoadEvent;
+import org.hibernate.event.spi.LoadEventListener;
+import org.hibernate.internal.CoreLogging;
+import org.hibernate.internal.CoreMessageLogger;
+import org.hibernate.loader.entity.CacheEntityLoaderHelper.EntityStatus;
+import org.hibernate.loader.entity.CacheEntityLoaderHelper.PersistenceContextEntry;
+import org.hibernate.persister.entity.EntityPersister;
+import org.hibernate.reactive.persister.entity.impl.ReactiveEntityPersister;
+
+/**
+ * @author Gavin King
+ *
+ * @see org.hibernate.loader.entity.CacheEntityLoaderHelper
+ */
+public class ReactiveCacheEntityLoaderHelper {
+
+	private static final CoreMessageLogger log = CoreLogging.messageLogger( ReactiveCacheEntityLoaderHelper.class );
+
+	private ReactiveCacheEntityLoaderHelper() {}
+
+	/**
+	 * Attempts to locate the entity in the session-level cache.
+	 * <p/>
+	 * If allowed to return nulls, then if the entity happens to be found in
+	 * the session cache, we check the entity type for proper handling
+	 * of entity hierarchies.
+	 * <p/>
+	 * If checkDeleted was set to true, then if the entity is found in the
+	 * session-level cache, it's current status within the session cache
+	 * is checked to see if it has previously been scheduled for deletion.
+	 *
+	 * @param event The load event
+	 * @param keyToLoad The EntityKey representing the entity to be loaded.
+	 * @param options The load options.
+	 *
+	 * @return The entity from the session-level cache, or null.
+	 *
+	 * @throws HibernateException Generally indicates problems applying a lock-mode.
+	 */
+	public static PersistenceContextEntry loadFromSessionCache(
+			final LoadEvent event,
+			final EntityKey keyToLoad,
+			final LoadEventListener.LoadType options) throws HibernateException {
+
+		SessionImplementor session = event.getSession();
+		Object old = session.getEntityUsingInterceptor( keyToLoad );
+
+		if ( old != null ) {
+			// this object was already loaded
+			EntityEntry oldEntry = session.getPersistenceContext().getEntry( old );
+			if ( options.isCheckDeleted() ) {
+				Status status = oldEntry.getStatus();
+				if ( status == Status.DELETED || status == Status.GONE ) {
+					log.debug(
+							"Load request found matching entity in context, but it is scheduled for removal; returning null" );
+					return new PersistenceContextEntry( old, EntityStatus.REMOVED_ENTITY_MARKER );
+				}
+			}
+			if ( options.isAllowNulls() ) {
+				final EntityPersister persister = event.getSession()
+						.getFactory()
+						.getEntityPersister( keyToLoad.getEntityName() );
+				if ( !persister.isInstance( old ) ) {
+					log.debug(
+							"Load request found matching entity in context, but the matched entity was of an inconsistent return type; returning null"
+					);
+					return new PersistenceContextEntry( old, EntityStatus.INCONSISTENT_RTN_CLASS_MARKER );
+				}
+			}
+			upgradeLock( old, oldEntry, event.getLockOptions(), event.getSession() );
+		}
+
+		return new PersistenceContextEntry( old, EntityStatus.MANAGED );
+	}
+
+	/**
+	 * see org.hibernate.event.internal.AbstractLockUpgradeEventListener#upgradeLock(Object, EntityEntry, LockOptions, EventSource)
+	 */
+	private static void upgradeLock(Object object, EntityEntry entry, LockOptions lockOptions, EventSource source) {
+
+		LockMode requestedLockMode = lockOptions.getLockMode();
+		if ( requestedLockMode.greaterThan( entry.getLockMode() ) ) {
+			// The user requested a "greater" (i.e. more restrictive) form of pessimistic lock
+
+			if ( entry.getStatus() != Status.MANAGED ) {
+				throw new ObjectDeletedException(
+						"attempted to lock a deleted instance",
+						entry.getId(),
+						entry.getPersister().getEntityName()
+				);
+			}
+
+			final EntityPersister persister = entry.getPersister();
+			final boolean cachingEnabled = persister.canWriteToCache();
+			SoftLock lock = null;
+			Object ck = null;
+			try {
+				if ( cachingEnabled ) {
+					EntityDataAccess cache = persister.getCacheAccessStrategy();
+					ck = cache.generateCacheKey( entry.getId(), persister, source.getFactory(), source.getTenantIdentifier() );
+					lock = cache.lockItem( source, ck, entry.getVersion() );
+				}
+
+				if ( persister.isVersioned() && requestedLockMode == LockMode.FORCE  ) {
+					// todo : should we check the current isolation mode explicitly?
+					Object nextVersion = persister.forceVersionIncrement( entry.getId(), entry.getVersion(), source );
+					entry.forceLocked( object, nextVersion );
+				}
+				else {
+					( (ReactiveEntityPersister) persister )
+							.lockReactive( entry.getId(), entry.getVersion(), object, lockOptions, source );
+				}
+				entry.setLockMode( requestedLockMode );
+			}
+			finally {
+				// the database now holds a lock + the object is flushed from the cache,
+				// so release the soft lock
+				if ( cachingEnabled ) {
+					persister.getCacheAccessStrategy().unlockItem( source, ck, lock );
+				}
+			}
+
+		}
+	}
+}

--- a/hibernate-reactive-core/src/test/java/org/hibernate/reactive/LockOnUpdateTest.java
+++ b/hibernate-reactive-core/src/test/java/org/hibernate/reactive/LockOnUpdateTest.java
@@ -1,0 +1,116 @@
+/* Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * Copyright: Red Hat Inc. and Hibernate Authors
+ */
+package org.hibernate.reactive;
+
+import java.util.Collection;
+import java.util.List;
+
+import javax.persistence.Basic;
+import javax.persistence.Entity;
+import javax.persistence.Id;
+import javax.persistence.Table;
+
+import org.hibernate.LockMode;
+
+import org.junit.Test;
+
+import io.smallrye.mutiny.Uni;
+import io.vertx.ext.unit.TestContext;
+
+/**
+ * @author Barry LaFond
+ */
+public class LockOnUpdateTest extends BaseReactiveTest {
+
+	@Override
+	protected Collection<Class<?>> annotatedEntities() {
+		return List.of( Record.class );
+	}
+
+	@Test
+	public void testLockDuringUpdate(TestContext context) {
+		Record secondRecord = new Record( "FIRST" );
+		secondRecord.text = "I'm the first record";
+		test(
+				context,
+				getMutinySessionFactory()
+						.withTransaction( session -> session.persist( secondRecord )
+								.call( session::flush )
+								.chain( () -> this.updateRecord( "FIRST", true ) )
+						)
+						.chain( () -> getMutinySessionFactory()
+								.withSession( session -> session.find( Record.class,
+										secondRecord.name
+								) ) )
+						.invoke( (record) -> context.assertTrue( record.text.contains( "updated" ) ) )
+		);
+	}
+
+	@Test
+	public void testLockBeforeUpdate(TestContext context) {
+		Record secondRecord = new Record( "SECOND" );
+		secondRecord.text = "I'm the second record";
+		test(
+				context,
+				getMutinySessionFactory()
+						.withTransaction( session -> session.persist( secondRecord )
+								.call( session::flush )
+								.call( () -> session.lock( secondRecord, LockMode.PESSIMISTIC_WRITE ) )
+								.chain( () -> this.updateRecord( "SECOND", false ) )
+						)
+						.chain( () -> getMutinySessionFactory()
+								.withSession( session -> session.find( Record.class,
+										secondRecord.name
+								) ) )
+						.invoke( (record) -> context.assertTrue( record.text.contains( "updated" ) ) )
+		);
+	}
+
+	private Uni<Record> updateRecord(final String name, final boolean doLock) {
+		if ( doLock ) {
+			return getMutinySessionFactory()
+					.withTransaction( session -> session.find( Record.class, name,
+							LockMode.PESSIMISTIC_WRITE
+					) )
+					.map( entity -> entity.setText( "I'm an updated record" ) );
+		}
+
+		return getMutinySessionFactory()
+				.withTransaction( session -> session.find( Record.class, name ) )
+				.map( entity -> entity.setText( "I'm an updated record" ) );
+
+	}
+
+	@Entity(name = "Record")
+	@Table(name = "FRecord")
+	static class Record {
+		@Id
+		String name;
+		@Basic(optional = false)
+		String text;
+
+		public Record() {
+		}
+
+		public Record(final String name) {
+			this.name = name;
+		}
+
+		public String getName() {
+			return name;
+		}
+
+		public Record setName(final String name) {
+			this.name = name;
+			return this;
+		}
+
+		public Record setText(final String text) {
+			this.text = text;
+			return this;
+		}
+	}
+}

